src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,122 @@
 package core.basesyntax;
+import java.util.Arrays;
+import java.util.NoSuchElementException;
 
 public class ArrayList<T> implements List<T> {
+    private Object[] arrayList;
+    private int size;
+    private final int defaultSize = 10;
+
+    public ArrayList() {
+        arrayList = new Object[defaultSize];
+        size = 0;
+    }
+
+    public void ensureCapacity() {
+        if (size == arrayList.length) {
+            arrayList = Arrays.copyOf(arrayList, arrayList.length + (arrayList.length >> 1));
+        }
+    }
+
+    public void ensureCapacity(int requiredCapacity) {
+        if (requiredCapacity > arrayList.length) {
+            int newCapacity = Math.max((int) (arrayList.length * 1.5), requiredCapacity);
+            arrayList = Arrays.copyOf(arrayList, newCapacity);
+        }
+    }
+
+    public void ensureIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
+    }
+
+    public void ensureIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
+    }
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        arrayList[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        ensureIndexForAdd(index);
+        ensureCapacity();
+        for (int i = size - 1; i >= index; i--) {
+            arrayList[i + 1] = arrayList[i];
+        }
+        arrayList[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        ensureCapacity(size + list.size());
```

Comment:
Ensure that the array resizing logic is encapsulated in a separate method to maintain single responsibility .